File: coffee-shop-app/backend/db_connection.py
import os
from dotenv import load_dotenv
import psycopg2

# Muat variabel dari file .env
load_dotenv()

def connect_to_postgres():
    conn = None
    try:
        conn = psycopg2.connect(
            host=os.getenv("DB_HOST"),
            database=os.getenv("DB_DATABASE"),
            user=os.getenv("DB_USER"),
            password=os.getenv("DB_PASSWORD")
        )
        print("✅ Berhasil terhubung ke database PostgreSQL!")
        return conn

    except (Exception, psycopg2.DatabaseError) as error:
        print(f"❌ Error saat terhubung ke PostgreSQL: {error}")
        return conn # Akan mengembalikan None jika error

# --- Cara Menggunakan Fungsi ---
if __name__ == "__main__":
    # Panggil fungsi untuk mendapatkan koneksi
    connection = connect_to_postgres()

    # Periksa apakah koneksi berhasil sebelum menjalankan query
    if connection is not None:
        try:
            # Buat objek cursor untuk mengeksekusi perintah SQL
            cursor = connection.cursor()
            
            # Contoh query sederhana: cek versi PostgreSQL
            print("\nMenjalankan query untuk mengecek versi...")
            cursor.execute("SELECT version();")
            
            # Ambil satu hasil dari query
            db_version = cursor.fetchone()
            print(f"Versi PostgreSQL: {db_version[0]}")

            # Tutup cursor
            cursor.close()
            
        except (Exception, psycopg2.DatabaseError) as error:
            print(f"❌ Error saat menjalankan query: {error}")
            
        finally:
            # PASTIKAN KONEKSI DITUTUP setelah selesai
            if connection is not None:
                connection.close()
                print("Koneksi ke database telah ditutup.")

File: coffee-shop-app/frontend/db_connection.js
require('dotenv').config(); // Muat variabel dari file .env
const { Client } = require('pg');

async function connectToPostgres() {
    const client = new Client({
        host: process.env.DB_HOST,
        user: process.env.DB_USER,
        port: process.env.DB_PORT,
        password: process.env.DB_PASSWORD,
        database: process.env.DB_DATABASE
    });

    try {
        await client.connect();
        console.log("✅ Berhasil terhubung ke database PostgreSQL!");
        return client;
    } catch (error) {
        console.error(`❌ Error saat terhubung ke PostgreSQL: ${error.stack}`);
        return null; // Mengembalikan null jika error
    }
}

// --- Cara Menggunakan Fungsi ---
async function main() {
    // Panggil fungsi untuk mendapatkan client
    const client = await connectToPostgres();

    // Periksa apakah koneksi berhasil sebelum menjalankan query
    if (client) {
        try {
            console.log("\nMenjalankan query untuk mengecek versi...");
            const res = await client.query('SELECT version();');
            
            console.log(`Versi PostgreSQL: ${res.rows[0].version}`);

        } catch (error) {
            console.error(`❌ Error saat menjalankan query: ${error.stack}`);
        } finally {
            // PASTIKAN KONEKSI DITUTUP setelah selesai
            await client.end();
            console.log("Koneksi ke database telah ditutup.");
        }
    }
}

// Jalankan fungsi utama
main();